crates/common/tedge_utils/src/notify.rs

```diff
@@ -0,0 +1,393 @@
+use std::{
+    collections::{HashMap, HashSet},
+    path::{Path, PathBuf},
+};
+
+use async_stream::try_stream;
+use futures::Stream;
+use notify::{
+    event::{AccessKind, AccessMode, CreateKind, DataChange, ModifyKind, RemoveKind},
+    Config, EventKind, INotifyWatcher, RecommendedWatcher, RecursiveMode, Watcher,
+};
+use tokio::sync::mpsc::{channel, Receiver};
+use try_traits::default::TryDefault;
+
+use crate::fs_notify::{FileEvent, NotifyStreamError};
+
+type DirPath = PathBuf;
+type MaybeFileName = Option<String>;
+type Metadata = HashMap<DirPath, HashMap<MaybeFileName, HashSet<FileEvent>>>;
+
+#[derive(Debug)]
+struct WatchMetadata {
+    metadata: Metadata,
+}
+
+impl WatchMetadata {
+    fn has_event_for(&self, dir_path: &Path, file_name: String, event: &FileEvent) -> bool {
+        if self.metadata.contains_key(dir_path) {
+            let maybe_file_name_entry = self.metadata.get(dir_path).unwrap(); // safe right?
+
+            if maybe_file_name_entry.contains_key(&Some(file_name.clone())) {
+                let events = maybe_file_name_entry.get(&Some(file_name)).unwrap();
+                return events.contains(event);
+            } else if maybe_file_name_entry.contains_key(&None) {
+                let events = maybe_file_name_entry.get(&None).unwrap();
+                return events.contains(event);
+            }
+        }
+        false
+    }
+}
+
+pub struct NotifyStream {
+    watcher: INotifyWatcher,
+    pub rx: Receiver<(PathBuf, FileEvent)>,
+    metadata: WatchMetadata,
```

Comment:
Why not simply remove one level?

```suggestion
    metadata: Metadata,
```